### PR TITLE
Improved type hierarchy for type mismatches 

### PR DESCRIPTION
**Pull request by MekhrubonT**
This PR tries to improve the types hierarchy for type mismatches.

I've created abstractions for variants, records, constructors, fields and extension constructors types mismatch. Also fixed some existing error output.

### CODE REVIEW

**Comment by gasche** [comment]
Note: this was previously part of #8821 -- the present PR is improved thanks to the review comments over there.

File: typing/includecore.ml [hunk 1]
```diff
@@ -122,134 +122,225 @@ let type_manifest env ty1 params1 ty2 params2 priv2 =
...
+let report_constructor_mismatch first second decl ppf err =
+  let pr fmt  = Format.fprintf ppf fmt in
+  match (err : constructor_mismatch) with
+  | Type s -> pr "The types for constructor %s are not equal." (Ident.name s)
+  | Arity s -> pr "The arities for constructor %s differ." (Ident.name s)
+  | Inline_record err -> report_record_mismatch first second decl ppf err
+  | Explicit_return_type ord ->
+      pr "%s has explicit return type and %s doesn't."
+        (String.capitalize_ascii (choose ord first second))
+        (choose (Ctype.Unification_trace.swap_position ord) first second)
```
Comment (gasche) [inline typing/includecore.ml]: Minor nitpick: I would define an auxiliary `choose_other` function for this.

File: typing/includecore.ml [hunk 2]
```diff
@@ -122,134 +122,225 @@ let type_manifest env ty1 params1 ty2 params2 priv2 =
...
+type position = Ctype.Unification_trace.position = First | Second
+
+let choose ord first second =
+  match ord with
+  | First -> first
+  | Second -> second
+
+type label_mismatch =
+  | Type of Ident.t
+  | Mutable of Ident.t
```
Comment (gasche) [inline typing/includecore.ml]: I think that `Mutability` would be more in line with the general naming convention of those constructors.)

Comment (Octachron) [inline typing/includecore.ml]: Maybe more `Mutability` rather than `Mutable` to have a noun in both cases.

File: typing/includecore.ml [hunk 3]
```diff
@@ -122,134 +122,225 @@ let type_manifest env ty1 params1 ty2 params2 priv2 =
...
+  | Inline_record of record_mismatch
+  | Explicit_return_type of position
+
+type variant_mismatch =
+  | Constructor_mismatch of constructor_mismatch
+  | Constructor_names of int * Ident.t * Ident.t
+  | Constructor_missing of position * Ident.t
+
+type extension_constructor_mismatch =
+  | Privacy
```
Comment (Octachron) [inline typing/includecore.ml]: Maybe `Privacy_mismatch` to be better balanced with `Constructor_mismatch` below. Or the reverse?

File: typing/includecore.ml [hunk 4]
```diff
@@ -122,134 +122,225 @@ let type_manifest env ty1 params1 ty2 params2 priv2 =
...
-  | Field_missing of bool * Ident.t
-  | Record_representation of bool   (* true means second one is unboxed float *)
-  | Unboxed_representation of bool  (* true means second one is unboxed *)
+  | Record_mismatch of record_mismatch
+  | Variant_mismatch of variant_mismatch
+  | Extension_constructor_mismatch of extension_constructor_mismatch
+  | Unboxed_representation of position
   | Immediate
 
+let report_label_mismatch _first _second ppf err =
```
Comment (Octachron) [inline typing/includecore.ml]: If the fields are not used, could they be removed?

File: typing/includecore.ml [hunk 5]
```diff
@@ -69,7 +69,7 @@ let is_absrow env ty =
 let type_manifest env ty1 params1 ty2 params2 priv2 =
   let ty1' = Ctype.expand_head env ty1 and ty2' = Ctype.expand_head env ty2 in
   match ty1'.desc, ty2'.desc with
-    Tvariant row1, Tvariant row2 when is_absrow env (Btype.row_more row2) ->
+  | Tvariant row1, Tvariant row2 when is_absrow env (Btype.row_more row2) ->
```
Comment (Octachron) [inline typing/includecore.ml]: The three changes above are unrelated to the current PR.

File: typing/includecore.ml (hunk 1 above)
Comment (Octachron) [inline typing/includecore.ml]: Why not just `choose ord second first`?

File: typing/includecore.ml [hunk 6]
```diff
@@ -122,134 +122,225 @@ let type_manifest env ty1 params1 ty2 params2 priv2 =
...
-let type_declarations ?(equality = false) ~loc env ~mark name decl1 path decl2 =
+let compare_records_with_representation ~loc env params1 params2 n
+      labels1 labels2 rep1 rep2
+  =
+  match compare_records ~loc env params1 params2 n labels1 labels2 with
+  | None when rep1 <> rep2 ->
+    Some (
+      Unboxed_representation
+            (if rep2 = Record_float then Second else First) : record_mismatch
+    )
```
Comment (Octachron) [inline typing/includecore.ml]: This is a lot of space for two nested constructor. Could we have

```OCaml
  let pos = if rep2 = Record_float then Second else First in
  Some ( Unboxed_float_representation pos  )
```
?

File: typing/includecore.ml [hunk 7]
```diff
@@ -364,22 +456,24 @@ let extension_constructors ~loc env ~mark id ext1 ext2 =
...
+      | Some r1, Some r2 ->
+          if Ctype.equal env true [r1] [r2] then
+            compare_constructor_arguments ~loc env id [r1] [r2]
+              ext1.ext_args ext2.ext_args
+        else Some (Type id : constructor_mismatch)
+      | Some _, None -> Some (Explicit_return_type First)
+      | None, Some _ -> Some (Explicit_return_type Second)
+      | None, None ->
+          compare_constructor_arguments ~loc env id
+            ext1.ext_type_params ext2.ext_type_params
```
Comment (Octachron) [inline typing/includecore.ml]: This code looks a lot like `compare_ext_constructors`, could we group it together with `compare_constructors`?

File: typing/includecore.ml [hunk 8]
```diff
@@ -122,134 +122,225 @@ let type_manifest env ty1 params1 ty2 params2 priv2 =
...
+        "uses unboxed float representation"
+
+let report_constructor_mismatch first second decl ppf err =
+  let pr fmt  = Format.fprintf ppf fmt in
+  match (err : constructor_mismatch) with
+  | Type s -> pr "The types for constructor %s are not equal." (Ident.name s)
+  | Arity s -> pr "The arities for constructor %s differ." (Ident.name s)
+  | Inline_record err -> report_record_mismatch first second decl ppf err
+  | Explicit_return_type ord ->
+      pr "%s has explicit return type and %s doesn't."
```
Comment (Octachron) [inline typing/includecore.ml:203]: I think that the error message should focus on GADTs rather than explicit return types. The fact that GADTs are enabled by having an explicit return type is more of an implementation detail.  Also it would be nice to test this new error message.

File: typing/includecore.ml [hunk 9]
```diff
@@ -122,134 +122,225 @@ let type_manifest env ty1 params1 ty2 params2 priv2 =
...
+        (String.capitalize_ascii (choose ord first second))
+        (choose (Ctype.Unification_trace.swap_position ord) first second)
+
+let report_variant_mismatch first second decl ppf err =
+  let pr fmt = Format.fprintf ppf fmt in
+  match (err : variant_mismatch) with
+  | Constructor_mismatch err ->
+    report_constructor_mismatch first second decl ppf err
+  | Constructor_names (n, name1, name2) ->
+      pr "Fields number %i have different names, %s and %s."
```
Comment (Octachron) [inline typing/includecore.ml]: `Fields number` → `Constructors number`

File: typing/includecore.ml [hunk 10]
```diff
@@ -122,134 +122,225 @@ let type_manifest env ty1 params1 ty2 params2 priv2 =
...
+        n (Ident.name name1) (Ident.name name2)
+  | Constructor_missing (ord, s) ->
+      pr "The field %s is only present in %s %s."
```
Comment (Octachron) [inline typing/includecore.ml]: `field` → `constructor`

File: typing/includecore.ml [hunk 11]
```diff
@@ -122,134 +122,225 @@ let type_manifest env ty1 params1 ty2 params2 priv2 =
...
+
+type label_mismatch =
+  | Type of Ident.t
+  | Mutable of Ident.t
+
+type record_mismatch =
+  | Label_mismatch of label_mismatch
+  | Label_names of int * Ident.t * Ident.t
+  | Label_missing of position * Ident.t
+  | Unboxed_representation of position
```
Comment (Octachron) [inline typing/includecore.ml]: This need to be `Unboxed_float_representation` otherwise it is too easy to forget that this error and the generic `Unboxed_representation` represent different issues.

**Review by Octachron — COMMENTED** [review]
The finer-grained types for type mismatches error message look nice. I think some of the constructor names needs some more discussion, and it would be nice to have a test for the new GADT/non-GADT mismatch error.

File: typing/includecore.ml (hunk 1 above)
Comment (MekhrubonT) [reply typing/includecore.ml]: It's easier to see the difference

**Review by gasche — COMMENTED** [review]
(I just found that I had two comments lying around from yesterday, mostly mirroring @Octachron's excellent remarks.)

File: typing/includecore.ml (hunk 3 above)
Comment (MekhrubonT) [reply typing/includecore.ml]: I would suggest to call it Constructor_privacy to be consistent with record_mismatch and variant_mismatch. It seems to me that suffix "_mismatch" in constructors' names is indicating type mismatch in deeper level of expression.

File: typing/includecore.ml (hunk 8 above)
Comment (lpw25) [reply typing/includecore.ml:203]: I disagree with this. Giving an explicit return type does not make something a GADT and the error is caused by giving an explicit return type on one side and not on the other. Things like:
```ocaml
module M : sig type t = T of int end = struct type t = T : int -> t end
```
get this error and do not involve any GADTs. The term "explicit return type" is used in the manual to describe this syntax, so it seems the correct term to use in this error message.

File: typing/includecore.ml (hunk 1 above)
Comment (MekhrubonT) [reply typing/includecore.ml]: I don't actually see the profit of creating such a function.

Comment (gasche) [reply typing/includecore.ml]: The caller code would read more nicely.

File: typing/includecore.ml (hunk 8 above)
Comment (Octachron) [reply typing/includecore.ml:203]: Isn't the detection of GADTs an entirely syntactic approximation based on the existence of an explicit return type right now?
But I see your point, if the error is tied to a syntactic approximation, let's keep the error message at the syntax level.

Comment (lpw25) [reply typing/includecore.ml:203]: > Isn't the detection of GADTs an entirely syntactic approximation based on the existence of an explicit return type right now?

Mostly, although weirdly not entirely -- for example the variance code tries to detect whether the constructor is actually a GADT.

Comment (gasche) [reply typing/includecore.ml:203]: I find this discussion a bit awkward given that it is perfectly sensible to define the term GADT to *include* standard algebraic datatypes (usually when we say that something is a "generalized foo" we include the foos in that), and that even if we usually use the term to say "a type that actually requires the generalized features", it is unclear whether people mean that existential types are used, or type equalities, or both.

(I'm perfectly happy with keeping things as they are.)

File: typing/includecore.ml [hunk 12]
```diff
@@ -198,24 +201,24 @@ let report_constructor_mismatch first second decl ppf err =
...
   | Constructor_missing (ord, s) ->
-      pr "The field %s is only present in %s %s."
+      pr "The constructor %s is only present in %s %s."
         (Ident.name s) (choose ord first second) decl
 
 let report_extension_constructor_mismatch first second decl ppf err =
   let pr fmt = Format.fprintf ppf fmt in
   match (err : extension_constructor_mismatch) with
-  | Privacy -> pr "A private type would be revealed"
+  | Constructor_privacy -> pr ".A private type would be revealed"
```
Comment (gasche) [inline typing/includecore.ml]: Why is there a new dot at the beginning of the message here? (Just a typo?)

File: typing/includecore.ml [hunk 13]
```diff
@@ -364,22 +457,15 @@ let extension_constructors ~loc env ~mark id ext1 ext2 =
...
+    let r =
+      compare_constructors ~loc env id ext1.ext_type_params ext2.ext_type_params
+        ext1.ext_ret_type ext2.ext_ret_type
+        ext1.ext_args ext2.ext_args
+    in
+    match r with
+    | Some r -> Some (Constructor_mismatch r)
+    | None -> match ext1.ext_private, ext2.ext_private with
+        Private, Public -> Some Constructor_privacy
+      | _, _ -> None
```
Comment (gasche) [inline typing/includecore.ml:470]: @lpw25: this factorization (and also `compare_variants`) seems strongly related to your recent commit fb497f1c8247914c6610c2e26b6abd72874104f9. (If I understand correctly, it's a nicer way to ensure that the code correct.)

**Review by gasche — APPROVED** [review]
The changes in `includecore.ml` are not just related to error messages, they also make the code nicer. I believe that they are correct, but I would welcome another review -- this part touches the soundness of the type-checker, not just the behavior in the error case.

File: typing/includecore.ml (hunk 12 above)
Comment (MekhrubonT) [reply typing/includecore.ml]: Yes, it's a typo.

File: typing/includecore.ml [hunk 14]
```diff
@@ -122,134 +122,226 @@ let type_manifest env ty1 params1 ty2 params2 priv2 =
 
 (* Inclusion between type declarations *)
 
+type position = Ctype.Unification_trace.position = First | Second
+
+let choose ord first second =
+  match ord with
+  | First -> first
+  | Second -> second
+let choose_other ord = match ord with
```
Comment (lpw25) [inline typing/includecore.ml]: Please make the formatting here consistent. Also I'd prefer it if you eta-expanded `choose_other` but that's mostly subjective.

File: typing/includecore.ml [hunk 15]
```diff
@@ -122,134 +122,226 @@ let type_manifest env ty1 params1 ty2 params2 priv2 =
...
-  | Field_type of Ident.t
-  | Field_mutable of Ident.t
-  | Field_arity of Ident.t
-  | Field_names of int * Ident.t * Ident.t
-  | Field_missing of bool * Ident.t
-  | Record_representation of bool   (* true means second one is unboxed float *)
-  | Unboxed_representation of bool  (* true means second one is unboxed *)
+  | Record_mismatch of record_mismatch
+  | Variant_mismatch of variant_mismatch
+  | Extension_constructor_mismatch of extension_constructor_mismatch
```
Comment (lpw25) [inline typing/includecore.ml]: I think this constructor isn't actually used.

**Review by lpw25 — APPROVED** [review]
LGTM

File: typing/includecore.ml (hunk 13 above)
Comment (lpw25) [reply typing/includecore.ml:470]: It is related and is indeed a nicer fix.

**Comment by lpw25** [comment]
(Note that @MekhrubonT is currently interning at Jane St. so he doesn't need to sign a CLA as his contributions are already covered by our corporate CLA)
